pkg/proto/merge/merge.go

```diff
@@ -58,18 +70,40 @@ var ReplaceMergeFn MergeFunction = func(dst, src protoreflect.Message) {
 	})
 }
 
+// CatchAllVirtualHostCloneFn creates a clone if src is a catchall virtual host.
+var CatchAllVirtualHostCloneFn CloneFunction = func(dst, src protoreflect.Message) protoreflect.ProtoMessage {
+	catchall := false
+	dst.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+		if v.String() == "allow_any" {
+			catchall = true
+			return false
+		}
+		return true
+	})
+	// If it is catchall virtual host, clone and merge so that cached value does not get modified.
+	if catchall {
+		return proto.Clone(dst.Interface())
+	}
+	return dst.Interface()
+}
+
 var options = []OptionFn{
 	// Workaround https://github.com/golang/protobuf/issues/1359, merge duration properly
 	MergeFunctionOptionFn((&durationpb.Duration{}).ProtoReflect().Descriptor().FullName(), ReplaceMergeFn),
+	// We cache the catch all virtual host for performance reasons. This helps not to mutate cached object.
+	CloneFunctionOptionFn((&route.VirtualHost{}).ProtoReflect().Descriptor().FullName(), CatchAllVirtualHostCloneFn),
```

Comment:
Is there any way we can mark objects at creation that they are "must clone"? Then everything cached can be like

```
var cachedConfig = merge.MustClone(&SomeObject{...})
```